rfcs/0001-reason-message-diff.md

```diff
@@ -0,0 +1,53 @@
+### Context:
+
+When kpack produces new builds today, a Reason is given which may be one of five values: COMMIT, TRIGGER, CONFIG, STACK, BUILDPACK
+
+These one-word values are helpful for providing a basic level of context on why a new build has been created, but they don't provide enough information to inform users' decisions about what to do with the resulting build. They may be asking themselves: "Does this new build have some update that affects my app? Should I run my regression tests on it? Should I promote it to prod?" 
+
+Providing more information would reduce the friction in users taking action on new builds and increase customers' app security by making sure more patched builds make it into production faster. 
+
+### Goals:
+
+Propose an additional attribute, “Reason-Message” be added to the build object, which can be printed in the context of the existing “Reason” attribute. Contextual information can be provided in the Reason-Message and a diff may also be printed as appropriate. This “Reason-Message” would be printed for these two commands:
+
+$ kp build status image-name
+$ kp image status image-name
+
+The format in which it would be printed:
+
+Reason: [Reason-Code] - [Reason-Message]
+
+Examples: 
+**For Reason "COMMIT"**
+- Ex. Reason: COMMIT - Received a new commit with ID [truncated-SH] from branch [branch-name]. 
+
+**For Reason "CONFIG"**
+- Ex. Reason: CONFIG - The [image-name] image config was updated. See diff below.
+[output of diffing library on the two image config files]
+
+**For Reason "BUILDPACK"**
+- Ex. Reason: BUILDER - The buildpacks used by your app have changed. See diff below. 
+[output of diffing library on the two Bill of Materials' for the builds]
+
+**For Reason "STACK"**
+- Ex. Reason: STACK - Stack-name changed from dockerhub/sha1 to dockerhub/sha2
+
+**For Reason "TRIGGER"**
+- Ex. Reason: TRIGGER - A new build was manually triggered by [user] at [time]
+
+### Prior Art:
+We have been thinking about different ways to surface the [Bill of Materials](https://github.com/buildpacks/pack/issues/378) for a build and how to show the delta between builds for a few months now. This is an evolution of that work. Although the `pack` cli renders the contents of a single BoM alone with the `inspect-image` command, I believe we don't need to provide a BoM alone, because the purpose of kpack is to iteratively build images, and therefore build BoMs should be placed in context as much as possible.
+
+### Complexity/Risks:
+How does the output render when there are two Reasons for a new build?
+
+How does the user compare two non-adjacent builds? ex. builds 17 and 20
```

Comment:
I'm curious about how common this use case is.  I've been assuming that the more common use case is to determine the diff between subsequent builds.  This is something that we could implement along the lines of Danny's comment above `kp image diff <build-A> <build-B>` as a follow along feature.  Or if it equally as desired as subsequent builds, it could become the primary interface for diffing builds.